Aspose.PDF for .NET 25.5 Release Notes. (#551)
Aspose/releases.aspose.com: content/en/pdf/net/release-notes/2025/aspose-pdf-for-net-25-05-release-notes.md

```csharp

// For complete examples and data files, visit https://github.com/aspose-pdf/Aspose.PDF-for-.NET
private static void ExtractCertificate()
{
    // The path to the documents directory
    var dataDir = RunExamples.GetDataDir_AsposePdf_SecuritySignatures();

    // Open PDF document
    using (var document = new Aspose.Pdf.Document(dataDir + "ExtractSignatureInfo.pdf"))
    {
        using (var signature = new Aspose.Pdf.Facades.PdfFileSignature(document))
        {
            // Get signature names
            var signatureNames = signature.GetSignatureNames();
            foreach (var signatureName in signatureNames)
            {
                // Extract certificate
                X509Certificate2 certificate;
                if (signature.TryExtractCertificate(signatureName, out certificate))
                {
                    Console.WriteLine(certificate.SerialNumber);
                }
            }
        }
    }
}
```

> Functionality to create ordered lists has been added. The intent of this technique is to create lists of related items using list elements appropriate for their purposes. PDF files containing lists are normally created or repaired using a tool for authoring PDF. When markup is used that visually formats items as a list but does not indicate the list relationship, users may have difficulty navigating the information.

Reviewer: ```suggestion
Functionality to create ordered lists has been added. This technique creates lists of related items using list elements appropriate for their purposes. PDF files containing lists are normally created or repaired using a tool for authoring PDF. When markup is used to visually format items as a list but does not indicate the list relationship, users may have difficulty navigating the information.
```

        li_3.AppendChild(lBody_3);
        lBody_3.AppendChild(nestedListDepth1);
        rootList.AppendChild(li_3);

        rootElement.AppendChild(rootList);

        // Save Tagged PDF Document
        document.Save(dataDir + "OrderedList_out.pdf");
    }
}
```

Starting from version 25.5, you can use an external certificate containing the public key that corresponds to the private key used to sign the document to verify the signature. This was made at the User's request for ID PDFNET-38869.

```csharp

// For complete examples and data files, visit https://github.com/aspose-pdf/Aspose.PDF-for-.NET
private static bool VerifyWithPublicKeyCertificate1(X509Certificate2 certificate, string inputPdfFilePath)
{
    // Create an instance of PdfFileSignature for working with signatures in the document
    using (var fileSign = new Aspose.Pdf.Facades.PdfFileSignature(inputPdfFilePath))
    {
        // Get a list of signatures
        IList<Aspose.Pdf.Facades.SignatureName> signatureNames = fileSign.GetSignatureNames();
        
        // Verify the signature with the given name.
        return fileSign.VerifySignature(signatureNames[0], certificate);
    }
}
```

> In some cases (e.g in case of static XFA form), conversion to AcroForm may suffer from disabled `NeedsRendering` flag that signals that form isn't dynamically rendered. To convert such forms and remove XFA form it's recommended to use `IgnoreNeedsRendering` property to count document as dynamically rendered and to calculate element conversion properly. This enhancement is logged as PDFNET-59659 in our issue tracker.

Reviewer: ```suggestion
In some cases (e.g, in the case of a static XFA form), conversion to AcroForm may suffer from the disabled `NeedsRendering` flag that signals that the form isn't dynamically rendered. To convert such forms and remove the XFA form, it's recommended to use the `IgnoreNeedsRendering` property to count the document as dynamically rendered and to calculate element conversion properly. This enhancement is logged as PDFNET-59659 in our issue tracker.
```

```csharp

// For complete examples and data files, please go to https://github.com/aspose-pdf/Aspose.PDF-for-.NET
private static void ConvertXFAFormWithIgnoreNeedsRendering()
{
    // The path to the documents directory.
    string dataDir = RunExamples.GetDataDir_AsposePdf_Forms();

    // Load dynamic XFA form
    using (var document = new Aspose.Pdf.Document(dataDir + "DynamicXFAToAcroForm.pdf")) 
    {
        //check if XFA is present & if rendering should be overwritten
        if (!doc.Form.NeedsRendering && doc.Form.HasXfa)
        {
            doc.Form.IgnoreNeedsRendering = true;
        }

        // Set the form fields type as standard AcroForm
        document.Form.Type = Aspose.Pdf.Forms.FormType.Standard;

        // Save the resultant PDF
        document.Save(dataDir + "Standard_AcroForm_out.pdf");
    }
}
```

> Sometimes when converting PDF to XPS, you may find that the fonts embedded in the PDF are not in a format suitable for XPS. At the same time, if a font with such a name is not found in the system, this leads to corrupted text. While we can't fix an embedded font during conversion without access to the original font file, we can replace it. Also, avoiding the use of embedded TrueType fonts can reduce conversion time. This enhancement is logged as PDFNET-59784 in our issue tracker.

Reviewer: ```suggestion
When converting PDF to XPS, you may find that the fonts embedded in the PDF are not in a format suitable for XPS. At the same time, if a font with such a name is not found in the system, this leads to corrupted text. While we can't fix an embedded font during conversion without access to the original font file, we can replace it. Also, avoiding the use of embedded TrueType fonts can reduce conversion time. This enhancement is logged as PDFNET-59784 in our issue tracker.
```
